Add an option to skip some tasks in the Physics

GiudGiud opened the pull request:
refs #30533

GiudGiud committed:
Add an option to skip some tasks in the Physics
refs #30533

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2880468/), step Docs: sync website on b426da1 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/30549/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2884239/), step Generate coverage on 5395531 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>083e81</th>
      <th colspan=4>#30549 539553</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.48%</td>
      <td>85.48%</td>
      <td>+0.00%</td>
      <td>96.23%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>112669</td>
      <td>112696</td>
      <td>+27</td>
      <td>51</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>19144</td>
      <td>19143</td>
      <td>-1</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/30549/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/30549/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/stochastic_tools)
- [`subchannel`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/subchannel)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/30549/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud commented on framework/src/physics/PhysicsBase.C:
```diff
@@ -29,17 +29,17 @@ PhysicsBase::validParams()
       "that system is used for all solver variables.");
-
-  MooseEnum pc_options("default none", "none");
+  MooseEnum pc_options("default none no_additional", "no_additional");
```
i ll progressively remove none from use

lindsayad replied:
What does `no_additional` mean?

lindsayad replied:
when could you add additional preconditioning? We don't support `-pc_type composite`

lindsayad replied:
If this new option is supposed to represent deferring to some other physics, I think we should come up with a different name. I think something as verbose as `defer_to_other_physics` is fine/good

GiudGiud replied:
no_additional is the same intent as "none", the Physics does not set the preconditioning.
"none" is not accurate because there is a default preconditioning, and because the user might be passing petsc_options

lindsayad replied:
I think `defer_to_other_physics` is a better enumerator name

GiudGiud replied:
done

GiudGiud commented:
@lindsayad I think this could help with getting people to use Physics. The flexibility to go back to the MOOSE regular syntax in the same input file (rather than Problem/type=DumpObjectsProblem that lets you make another input instead)

lindsayad commented:
Yea I think being able to mix a portion of the Physics object creation and manual user object creation sounds potentially handy!

GiudGiud committed:
Address Alex's review:
- instead of 'none' for preconditioing, 'defer' which does not convey the impression
that no PC will be used

GiudGiud commented on framework/src/physics/PhysicsBase.C:
```diff
@@ -29,17 +29,17 @@ PhysicsBase::validParams()
       "that system is used for all solver variables.");
-
-  MooseEnum pc_options("default none", "none");
+  MooseEnum pc_options("default defer_to_other_physics", "defer_to_other_physics");
```
can try to deprecate none already actually since my TMAP pr is not merged

lindsayad replied:
I don't think `none` is so bad. It describes a lot of the current use cases of physics right? There's not a single input with multiple physics merged into the MOOSE herd is there?

lindsayad replied:
If I'm right about the last question, then right now having the default enum be `defer_to_other_physics` doesn't seem like the best choice since it doesn't conceptually describe any of our inputs

GiudGiud replied:
I can re-instate none later then haha. We can have as many options as we want to not add the preconditioning. 

>  There's not a single input with multiple physics merged into the MOOSE herd is there?

nope that s what the other PR enables though

GiudGiud replied:
> it doesn't conceptually describe any of our inputs

it's transparent though. "defer_to_other..." will always mean "this object is not taking care of preconditioning"

lindsayad replied:
well now I'm thinking just `defer` is the best name. Because I assume in a lot of physics inputs we are deferring to the `Preconditioning` block

lindsayad replied:
If you go with plain `defer`, then I think dropping `none` makes total sense

GiudGiud replied:
I ll do this now

GiudGiud commented:
nice thanks for the review!
I think this can help adoption for Physics. We'll see
